### DISCUSSION TITLE
Invoke async code in `ResponseError` trait method `error_response`

### DISCUSSION THREAD

**Discussion by mellowagain** [actix/actix-web#2593] [Q&A] [answered]
In order to handle errors in my route, I created a custom `struct` which implements the `ResponseError` trait of actix-web.

```rust
struct CustomError {
     // ...
}
```

On error, I'd like to invoke a async function:

```rust
impl ResponseError for CustomError {
    fn error_response(&self) -> HttpResponse {
        render_template("error.html").await // <-- Not possible to call .await in a sync function
    }

    // ...
}
```

Sadly Rust doesn't have async traits (without the `async_trait` crate). I've already tried spawning a new Runtime but that obviously doesn't work as this method gets called from within the existing Runtime. 

In `tokio` there is a `Handle` struct which allows one to retrieve a handle to the current runtime in which I could call `.block_on` for the async method, but I don't seem to find a way to acquire something similiar for the actix-web rt.

Any idea how to handle this?

**Comment by aliemjay** [comment, marked as answer]
I recommend rendering a simple response in `ResponseError` and then using a middleware for fully rendering the error page. With this approach you have the luxury of async as well as having HttpRequest.

In the middleware, you can downcast `HttpResponse::error()` to get the underlying error. You also have the option to pass arbitrary data from `ResponseError::error_response()` to the middleware using `HttpResponse::extensions()`.

A related example <https://github.com/svenstaro/miniserve/blob/f40e7d4c66775c8f89564ee1b2bbe4b2d7cfc660/src/errors.rs#L112>

**Reply by mellowagain** [reply]
Thank you very much, this does work perfectly.

For anyone else stumbling upon this, [here](https://github.com/mellowagain/gitarena/blob/0a6aa55/src/error.rs#L215-L259) is my implementation of the answer provided by this comment.
